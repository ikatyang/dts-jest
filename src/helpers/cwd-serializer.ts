export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
