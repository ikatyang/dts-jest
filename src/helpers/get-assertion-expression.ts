import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
