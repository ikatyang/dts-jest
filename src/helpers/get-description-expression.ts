import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
