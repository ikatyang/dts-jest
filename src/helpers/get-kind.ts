import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
