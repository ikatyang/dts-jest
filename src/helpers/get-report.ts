import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
