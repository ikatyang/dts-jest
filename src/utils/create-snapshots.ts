import * as ts from 'typescript';
import { Snapshot } from '../definitions';
import { traverse_node } from './traverse-node';

export const create_snapshots = (
  program: ts.Program,
  source_filename: string,
  lines: number[],
  flag: ts.TypeFormatFlags,
) => {
  const source_file = program.getSourceFile(source_filename);
  const snapshots: { [line: number]: Snapshot } = {};

  const rest_lines = lines.slice();

  const diagnostics = ts.getPreEmitDiagnostics(program, source_file);
  for (const diagnostic of diagnostics) {
    // tslint:disable-next-line:no-unnecessary-type-assertion
    const position = diagnostic.start!;
    const { line: error_line } = source_file.getLineAndCharacterOfPosition(
      position,
    );
    const trigger_line = error_line - 1;

    const line_index = rest_lines.indexOf(trigger_line);
    if (line_index !== -1) {
      snapshots[trigger_line] = {
        diagnostic: ts.flattenDiagnosticMessageText(
          diagnostic.messageText,
          '\n',
        ),
      };
      rest_lines.splice(line_index, 1);
    }
  }

  const checker = program.getTypeChecker();
  traverse_node(source_file, node => {
    const position = node.getStart(source_file);
    const { line: expression_line } = source_file.getLineAndCharacterOfPosition(
      position,
    );
    const trigger_line = expression_line - 1;

    const line_index = rest_lines.indexOf(trigger_line);
    if (line_index !== -1) {
      const target_node = node.getChildAt(0);
      const type = checker.getTypeAtLocation(target_node);
      snapshots[trigger_line] = {
        inference: checker.typeToString(type, node, flag),
      };
      rest_lines.splice(line_index, 1);
    }
  });

  return snapshots;
};
