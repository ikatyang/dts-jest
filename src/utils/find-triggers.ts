import * as _ts from 'typescript';
import { Trigger, TriggerFooter } from '../definitions';
import { find_trigger_bodies } from './find-trigger-bodies';
import { find_trigger_footers } from './find-trigger-footers';
import { find_trigger_headers } from './find-trigger-headers';
import { get_trigger_body_end_line } from './get-trigger-line';

export const find_triggers = (source_file: _ts.SourceFile, ts: typeof _ts) => {
  const triggers: Trigger[] = [];

  const headers = find_trigger_headers(source_file, ts);
  const bodies = find_trigger_bodies(source_file, headers, ts);
  const footers = find_trigger_footers(source_file, bodies, ts);

  for (let i = 0; i < headers.length; i++) {
    const header = headers[i];
    const body = bodies[i];
    let footer: TriggerFooter | undefined;

    const { line: body_end_line } = source_file.getLineAndCharacterOfPosition(
      body.end,
    );

    if (
      footers.length !== 0 &&
      body_end_line === get_trigger_body_end_line(footers[0].line)
    ) {
      footer = footers.shift();
    }

    triggers.push({
      header,
      body,
      footer,
    });
  }

  return triggers;
};
