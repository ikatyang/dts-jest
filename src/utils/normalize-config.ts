import { NormalizedConfig, RawConfig } from '../definitions';
import { load_compiler_options } from './load-compiler-options';
import { load_typescript } from './load-typescript';

export const normalize_config = (
  raw_config: RawConfig = {},
  root_dir: string,
): NormalizedConfig => {
  // istanbul ignore next
  const {
    test_type = true,
    test_value = false,
    compiler_options: raw_compiler_options = {},
    enclosing_declaration = false,
    typescript: typescript_id = 'typescript',
  } = raw_config;

  const { typescript, typescript_path } = load_typescript(
    replace_root_dir(typescript_id, root_dir),
  );

  // istanbul ignore next
  const {
    type_format_flags = typescript.TypeFormatFlags.NoTruncation,
  } = raw_config;

  const compiler_options = load_compiler_options(
    // istanbul ignore next
    typeof raw_compiler_options === 'string'
      ? replace_root_dir(raw_compiler_options, root_dir)
      : raw_compiler_options,
    root_dir,
    typescript,
  );

  return {
    test_type,
    test_value,
    compiler_options,
    type_format_flags,
    enclosing_declaration,
    typescript,
    typescript_path,
  };
};

function replace_root_dir(str: string, root_dir: string) {
  return str.replace('<rootDir>', root_dir);
}
