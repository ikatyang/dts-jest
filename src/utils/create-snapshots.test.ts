import * as ts from 'typescript';
import { cwd_serializer } from '../helpers/cwd-serializer';
import {
  get_fixture_filename,
  load_fixture_source_file,
} from '../helpers/load-fixture';
import { create_snapshots } from './create-snapshots';
import { find_triggers } from './find-triggers';
import { normalize_config } from './normalize-config';

expect.addSnapshotSerializer(cwd_serializer);

it('should create correctly', () => {
  expect(get_snapshots('example')).toMatchSnapshot();
});

it('should return empty array if there is no assertion', () => {
  expect(get_snapshots('no-assertion')).toHaveLength(0);
});

it('should throw error of there are some unmatched diagnostics', () => {
  expect(() => get_snapshots('unmatched')).toThrowErrorMatchingSnapshot();
});

function get_snapshots(id: string) {
  const full_id = `create-snapshots/${id}.ts`;
  const filename = get_fixture_filename(full_id);
  const source_file = load_fixture_source_file(full_id, ts);
  const triggers = find_triggers(source_file, ts);
  const config = normalize_config({}, process.cwd());
  return create_snapshots(filename, triggers, config);
}
