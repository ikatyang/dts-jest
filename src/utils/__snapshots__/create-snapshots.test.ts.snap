// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create correctly 1`] = `
Array [
  Object {
    "diagnostic": "Argument of type '\\"123\\"' is not assignable to parameter of type 'number'.",
    "line": 7,
  },
  Object {
    "diagnostic": "Argument of type '{ c: number; }' is not assignable to parameter of type '{ a: 1; }'.
  Object literal may only specify known properties, and 'c' does not exist in type '{ a: 1; }'.",
    "line": 10,
  },
  Object {
    "inference": "number",
    "line": 1,
  },
  Object {
    "inference": "number",
    "line": 4,
  },
]
`;

exports[`should throw error of there are some unmatched diagnostics 1`] = `
"Unmatched diagnostic(s) detected:

  <cwd>/fixtures/create-snapshots/unmatched.ts:1 Argument of type '\\"str\\"' is not assignable to parameter of type 'number'."
`;
